src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,419 @@
+                         true, "number of classes must be greater than or "
+                         "equal to 0 (equal to 0 in case of unspecified.)");
+
+  // Delta must be positive.
+  RequireParamValue<double>("delta", [](double x) { return x >= 0.0; }, true,
+      "delta must be non-negative");
+
+  // Delta must be positive.
+  RequireParamValue<int>("epochs", [](int x) { return x > 0; }, true,
+      "epochs must be non-negative");
+
+  // These are the matrices we might use.
+  arma::mat trainingSet;
+  arma::Row<size_t> labels;
+  arma::Row<size_t> rawLabels;
+  arma::mat testSet;
+  arma::Row<size_t> predictions;
+  arma::Col<size_t> mappings;
+  size_t numClasses;
+
+  // Load data matrix.
+  if (CLI::HasParam("training"))
+    trainingSet = std::move(CLI::GetParam<arma::mat>("training"));
+
+  // Check if the labels are in a separate file.
+  if (CLI::HasParam("training") && CLI::HasParam("labels"))
+  {
+    rawLabels = std::move(CLI::GetParam<arma::Row<size_t>>("labels"));
+    if (trainingSet.n_cols != rawLabels.n_cols)
+    {
+      Log::Fatal << "The labels must have the same number of points as the "
+          << "training dataset." << endl;
+    }
+  }
+  else if (CLI::HasParam("training"))
+  {
+    // Checking the size of training data if no labels are passed.
+    if (trainingSet.n_rows < 2)
+    {
+      Log::Fatal << "Can't get labels from training data since it has less "
+          << "than 2 rows." << endl;
+    }
+
+    // The initial predictors for y, Nx1.
+    rawLabels = arma::conv_to<arma::Row<size_t>>::from(
+        trainingSet.row(trainingSet.n_rows - 1));
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  data::NormalizeLabels(rawLabels, labels, mappings);
+
+  // Load the model, if necessary.
+  LinearSVM<>* model;
+  if (CLI::HasParam("input_model"))
+  {
+    model = CLI::GetParam<LinearSVM<>*>("input_model");
+  }
+  else
+  {
+    numClasses = CLI::GetParam<int>("number_of_classes") == 0 ?
+        arma::max(labels) + 1 : CLI::GetParam<int>("number_of_classes");
+    model = new LinearSVM<>(0, numClasses);
+  }
+
+  // Now, do the training.
+  if (CLI::HasParam("training"))
+  {
+    model->Lambda() = lambda;
+    model->Delta() = delta;
+    numClasses = model->NumClasses();
+
+    if (optimizerType == "lbfgs")
+    {
+      model->FitIntercept() = intercept;
+
+      ens::L_BFGS lbfgsOpt;
+      lbfgsOpt.MaxIterations() = maxIterations;
+      lbfgsOpt.MinGradientNorm() = tolerance;
+
+      Log::Info << "Training model with L-BFGS optimizer." << endl;
+
+      // This will train the model.
+      model->Train(trainingSet, labels, numClasses, lbfgsOpt);
+    }
+    else if (optimizerType == "psgd")
+    {
+      const double step_size = CLI::GetParam<double>("step_size");
+      const bool shuffle = !CLI::HasParam("shuffle");
+      const size_t maxIt = epochs * trainingSet.n_cols;
+
+      ens::ConstantStep decayPolicy(step_size);
+
+      #ifndef HAS_OPENMP
```

Comment:
There shouldn't be a need for this check here; regardless of whether or not OpenMP is available, `ens::ParallelSGD` is still available and will still work (even with one thread).  It may be worthwhile to print a warning if OpenMP is not in use and parallel SGD was chosen, like `Log::Warn << "Using parallel SGD, but OpenMP support is not available!" << std::endl;` or something.

Reply:
Removed.